File: src/main/java/com/sleepkqq/sololeveling/ui/aop/LoggingAspect.java
package com.sleepkqq.sololeveling.ui.aop;

import lombok.extern.slf4j.Slf4j;
import org.springframework.util.StopWatch;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.springframework.stereotype.Component;

@Slf4j
@Aspect
@Component
public class LoggingAspect {

  @Pointcut("within(com.sleepkqq.sololeveling.ui.api..*)")
  public void apiMethods() {}

  @Around("apiMethods()")
  public Object logExecutionTime(ProceedingJoinPoint joinPoint) throws Throwable {
    var stopWatch = new StopWatch();

    stopWatch.start();
    var result = joinPoint.proceed();
    stopWatch.stop();

    log.info(
        "api '{}' executed in {} ms",
        joinPoint.getSignature().getName(),
        stopWatch.getTotalTimeMillis()
    );

    return result;
  }
}

File: solo-leveling-ui-service/src/main/kotlin/com.sleepkqq.sololeveling/ui/aop/LoggingAspect.kt
package com.sleepkqq.sololeveling.ui.aop

import com.sleepkqq.sololeveling.ui.service.TgAuthService
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.util.StopWatch

@Aspect
@Component
class LoggingAspect(
	private val tgAuthService: TgAuthService
) {

	private val log = LoggerFactory.getLogger(LoggingAspect::class.java)

	@Pointcut("within(com.sleepkqq.sololeveling.ui.api..*)")
	fun apiMethods() {

	}

	@Around("apiMethods()")
	@Throws(Throwable::class)
	fun logExecutionTime(joinPoint: ProceedingJoinPoint): Any {
		val username = tgAuthService.findCurrentUser()
			?.username
			?: "anon"

		val stopWatch = StopWatch()
		stopWatch.start()
		val result = joinPoint.proceed()
		stopWatch.stop()

		log.info(
			">> api '{}' executed by '{}' in {} ms",
			joinPoint.signature.name,
			username,
			stopWatch.totalTimeMillis
		)

		return result
	}
}